server/src/modules/users/application/use-cases/authorize-user/authorize-user-dto.ts

```diff
@@ -0,0 +1,33 @@
+import Joi from 'joi'
+import express from 'express'
+
+export const SUPPORTED_OPEN_ID_RESPONSE_TYPES = ['code']
+export const SUPPORTED_OPEN_ID_SCOPE = ['openid']
+
+export interface AuthorizeUserDTOParams {
+  client_id: string
+  scope: string
+  response_type: string
+  redirect_uri: string
+}
+
+export interface AuthorizeUserDTO {
+  req: express.Request
+  params: AuthorizeUserDTOParams
+}
+
+export const AuthorizeUserDTOParamsSchema = Joi.object<AuthorizeUserDTOParams>({
+  client_id: Joi.string().required(),
+  scope: Joi.string()
+    .valid(...SUPPORTED_OPEN_ID_SCOPE)
+    .required(),
+  response_type: Joi.string()
+    .valid(...SUPPORTED_OPEN_ID_RESPONSE_TYPES)
+    .required(),
+  redirect_uri: Joi.string().uri().required(),
+}).options({ abortEarly: false })
+
+export const AuthorizeUserDTOSchema = Joi.object<AuthorizeUserDTO>({
+  req: Joi.object().required(),
```

Comment:
The issue is that by narrowing the schema at this point, I wouldn't be able to redirect the user to /login if the user doesn't exist, like done in the use-case (we return a 400 for all schema errors currently). The overridable method you showed me earlier would let me change that. Is that something I should include here?

Reply:
Since https://github.com/loolabs/waterpark/pull/221 is merged now, feel free to copy over https://github.com/loolabs/waterpark/blob/main/server/src/shared/app/typed-controller.ts and its dependencies to try it out. A word of warning -- it uses Zod. You might be able to modify it to use Joi validation.